src/Microsoft.AspNetCore.Mvc.Core/Internal/ProblemDetailsFilter.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Core;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    public class ProblemDetailsFilter : IActionFilter
+    {
+        private readonly IErrorDescriptionFactory _factory;
+
+        public ProblemDetailsFilter(IErrorDescriptionFactory factory)
+        {
+            _factory = factory ?? throw new ArgumentNullException(nameof(factory));
+        }
+
+        public void OnActionExecuting(ActionExecutingContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (context.Result == null || !(context.Result is StatusCodeResult statusCodeResult))
+            {
+                return;
+            }
+
+            var isExactlyStatusCodeResult = context.Result.GetType() == typeof(StatusCodeResult);
+            ProblemDetails problemDetails;
+            switch (context.Result)
+            {
+                case NotFoundResult notFoundResult:
+                    problemDetails = GetNotFoundProblemDetails(notFoundResult);
+                    break;
+                case BadRequestResult badRequestResult:
+                    problemDetails = GetBadRequestProblemDetails(badRequestResult);
+                    break;
+                case UnsupportedMediaTypeResult unsupportedMediaTypeResult:
+                    problemDetails = GetUnsupportedMediaTypeProblemDetails(unsupportedMediaTypeResult);
+                    break;
+                case StatusCodeResult result when isExactlyStatusCodeResult && result.StatusCode == StatusCodes.Status400BadRequest:
+                    problemDetails = GetBadRequestProblemDetails(result);
+                    break;
+                case StatusCodeResult result when isExactlyStatusCodeResult && result.StatusCode == StatusCodes.Status404NotFound:
```

Comment:
overtly complicated code ftw. But I do get 💯 points

Reply:
This is ugly, but I think it gets the job done. I want to go with something like this for now to unblock us, even if it's super inelegant